samza-test/src/test/java/org/apache/samza/test/processor/TestZkLocalApplicationRunner.java

```diff
@@ -0,0 +1,397 @@
+
+    // Run stream applications.
+    applicationRunner1.run(streamApp1);
+    applicationRunner2.run(streamApp2);
+
+    // Wait for message processing to start in both the processors.
+    processedMessagesLatch1.await();
+    processedMessagesLatch2.await();
+
+    LocalApplicationRunner applicationRunner3 = new LocalApplicationRunner(new MapConfig(applicationConfig2));
+
+    // Create a stream app with same processor id as SP2 and run it. It should fail.
+    StreamApplication streamApp3 = new TestStreamApplication(inputKafkaTopic, outputKafkaTopic, null, null);
+    applicationRunner3.run(streamApp3);
+
+    // The following line should throw up by handling duplicate processorId registration.
+    applicationRunner3.waitForFinish();
+  }
+
+  // Depends upon SAMZA-1302
+  //@Test(expected = Exception.class)
+  public void shouldKillStreamAppWhenZooKeeperDiesBeforeLeaderReElection() throws InterruptedException {
+    // Create StreamApplications.
+    StreamApplication streamApp1 = new TestStreamApplication(inputKafkaTopic, outputKafkaTopic, null, null);
+    StreamApplication streamApp2 = new TestStreamApplication(inputKafkaTopic, outputKafkaTopic, null, null);
+
+    // Run stream applications.
+    applicationRunner1.run(streamApp1);
+    applicationRunner2.run(streamApp2);
+
+    applicationRunner1.kill(streamApp1);
+    applicationRunner1.waitForFinish();
+    assertEquals(ApplicationStatus.SuccessfulFinish, applicationRunner1.status(streamApp2));
+
+    // Kill zookeeper server before debounce time. JobModel regeneration and ZK fencing will fail with exception.
+    zookeeper().shutdown();
+
+    applicationRunner1 = new LocalApplicationRunner(applicationConfig1);
+    applicationRunner1.run(streamApp1);
+    // This line should throw exception since Zookeeper is unavailable.
+    applicationRunner1.waitForFinish();
+  }
+
+  // Depends upon SAMZA-1302
+  //@Test
+  public void testRollingUpgrade() throws Exception {
+    publishKafkaEvents(inputKafkaTopic, NUM_KAFKA_EVENTS, PROCESSOR_IDS[0]);
+
+    List<TestKafkaEvent> messagesProcessed = new ArrayList<>();
+    StreamApplicationCallback streamApplicationCallback = messagesProcessed::add;
+
+    // Create StreamApplication from configuration.
+    CountDownLatch processedMessagesLatch1 = new CountDownLatch(1);
+    CountDownLatch processedMessagesLatch2 = new CountDownLatch(1);
+
+    StreamApplication streamApp1 = new TestStreamApplication(inputKafkaTopic, outputKafkaTopic, processedMessagesLatch1, streamApplicationCallback);
+    StreamApplication streamApp2 = new TestStreamApplication(inputKafkaTopic, outputKafkaTopic, processedMessagesLatch2,null);
+
+    // Run stream application.
+    applicationRunner1.run(streamApp1);
+    applicationRunner2.run(streamApp2);
+
+    processedMessagesLatch1.await();
+    processedMessagesLatch2.await();
+
+    applicationRunner1.kill(streamApp1);
+    applicationRunner1.waitForFinish();
+
+    int lastProcessedMessageId = -1;
```

Comment:
Was present in local, didn't update it here(missed during cleanup).
Added assertions.